atc/auditor/auditor.go

```diff
@@ -0,0 +1,175 @@
+package auditor
+
+import (
+	"net/http"
+
+	"code.cloudfoundry.org/lager"
+	"github.com/concourse/concourse/atc"
+)
+
+//go:generate counterfeiter . Auditor
+
+
+func NewAuditor(
+	EnableBuildAuditLog bool,
+	EnableContainerAuditLog bool,
+	EnableJobAuditLog bool,
+	EnablePipelineAuditLog bool,
+	EnableResourceAuditLog bool,
+	EnableSystemAuditLog bool,
+	EnableTeamAuditLog bool,
+	EnableWorkerAuditLog bool,
+	EnableVolumeAuditLog bool,
+	logger lager.Logger,
+) *auditor {
+	return &auditor{
+		EnableBuildAuditLog:     EnableBuildAuditLog ,
+		EnableContainerAuditLog: EnableContainerAuditLog,
+		EnableJobAuditLog:       EnableJobAuditLog,
+		EnablePipelineAuditLog:  EnablePipelineAuditLog,
+		EnableResourceAuditLog:  EnableResourceAuditLog,
+		EnableSystemAuditLog:    EnableSystemAuditLog,
+		EnableTeamAuditLog:      EnableTeamAuditLog,
+		EnableWorkerAuditLog:    EnableWorkerAuditLog,
+		EnableVolumeAuditLog:    EnableVolumeAuditLog,
+		logger:                  logger,
+	}
+}
+
+
+type Auditor interface {
+	LogAction(action string, userName string, r *http.Request)
+	ValidateAction(action string) bool
```

Comment:
Does `ValidateAction` need to be part of this interface?

Also small nit, but can we rename `LogAction` to `Audit`.

Reply:
You know, you're right on ValidateAction, I'm really only using that locally.  Let me refactor that accordingly.  And yeah, totally fine to change the name of that method.

let me know if there anything else you would like me to respond to, and sorry it's required this much of an in depth review.

Reply:
@pivotal-jwinters In setting `ValidateAction` to not be a part of the interface, I wanted to know your team's policy regarding internal/external tests.  (as that method won't be public anymore) 

ie. In order to test ValidateAction, I would think to create one set of tests that is part of the `package auditor_test`, and a second set of tests that is part of `package auditor`

From what I could see, that's not a practice really followed anywhere else in the code base, so I wanted to ask: Do you prefer having only external tests? or am I ok to also build internal tests?

Reply:
In general we only test exported methods because anything else is just an implementation detail. 

From the perspective of your tests, the only thing you really care about is if I `Audit` this event does it show up in my logs. You are already implicitly testing the `ValidateAction` method in these tests. So further explicit tests are not required.

Reply:
To be honest, that was my thought as well. (I've just worked a variety of organizations that have different ideologies when it comes to unit tests, so I tend to go to the most extreme version of unit testing, and scale down from there).

So I just hoped to find what your team would prefer.  I'll remove the explicit tests on the `ValidateAction` method.